File: backend/app/taxonomy.py
"""Fonte única da taxonomia CIMI Leads.

Carrega backend/app/taxonomy.json na importação e deriva todas as listas de
validação. Nenhum outro arquivo deve hardcodar produtos, perfis ou tags de
interesse.

IMPORTANTE: shared/taxonomy.json é a fonte EDITÁVEL. Este arquivo é uma cópia
sincronizada — editar só shared/taxonomy.json e rodar scripts/sync_taxonomy.py
(o pre-commit hook check-taxonomy-sync bloqueia commit se divergir).
"""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

_TAXONOMY_PATH = Path(__file__).resolve().parent / "taxonomy.json"


def _load_taxonomy() -> dict[str, Any]:
    try:
        with open(_TAXONOMY_PATH, "r", encoding="utf-8") as f:
            return json.load(f)
    except FileNotFoundError:
        logger.error("Taxonomy file not found: %s", _TAXONOMY_PATH)
        raise RuntimeError("taxonomy.json is required") from None
    except json.JSONDecodeError as exc:
        logger.error("Invalid taxonomy JSON: %s", exc)
        raise RuntimeError("taxonomy.json is invalid") from exc


_raw = _load_taxonomy()

PRODUCTS: list[dict[str, Any]] = _raw["products"]
LEGACY_PROFILES: dict[str, list[dict[str, Any]]] = _raw.get("legacy_profiles", {})
INTEREST_TYPES: list[str] = _raw["interest_types"]

# ── Índices derivados para validação rápida ─────────────────────────────────

# product_key -> set de slugs válidos (inclui legacy)
_CLASSIFICATION_VALID: dict[str, set[str]] = {}
# product_key -> label
_PRODUCT_LABELS: dict[str, str] = {}
# product_key -> slug -> label
_PROFILE_LABELS: dict[str, dict[str, str]] = {}

for p in PRODUCTS:
    key = p["key"]
    slugs = {pr["slug"] for pr in p["profiles"]}
    _CLASSIFICATION_VALID[key] = slugs
    _PRODUCT_LABELS[key] = p["label"]
    _PROFILE_LABELS[key] = {pr["slug"]: pr["label"] for pr in p["profiles"]}

# Adiciona legacy nos validadores (mas não na UI)
for prod_key, profiles in LEGACY_PROFILES.items():
    legacy_slugs = {pr["slug"] for pr in profiles}
    _CLASSIFICATION_VALID.setdefault(prod_key, set()).update(legacy_slugs)
    # labels de legacy também são necessários para CSV/history
    if prod_key not in _PROFILE_LABELS:
        _PROFILE_LABELS[prod_key] = {}
    for pr in profiles:
        _PROFILE_LABELS[prod_key][pr["slug"]] = pr["label"]

# Conjunto de todas as tags de interesse aceitas
ALLOWED_TAGS_SET: set[str] = set(INTEREST_TYPES)


def is_valid_classification(tag: str) -> tuple[bool, str | None]:
    """Valida uma tag de classificação (produto:perfil).

    Retorna (True, None) se válida.
    Retorna (False, reason) se inválida, onde reason é uma mensagem para o usuário.
    """
    if ":" not in tag:
        return False, None  # não é classificação, é tag de interesse
    product_key, _, slug = tag.partition(":")
    if product_key not in _CLASSIFICATION_VALID:
        return False, f"produto '{product_key}' desconhecido na classificação '{tag}'"
    if slug not in _CLASSIFICATION_VALID[product_key]:
        return False, f"perfil '{slug}' inválido para '{product_key}'"
    return True, None


def get_product_label(key: str) -> str | None:
    return _PRODUCT_LABELS.get(key)


def get_profile_label(product_key: str, slug: str) -> str | None:
    return _PROFILE_LABELS.get(product_key, {}).get(slug)


def get_taxonomy_payload() -> dict[str, Any]:
    """Retorna o payload para GET /api/taxonomy (sem legacy na lista de produtos)."""
    return {
        "products": PRODUCTS,
        "legacy_profiles": LEGACY_PROFILES,
        "interest_types": INTEREST_TYPES,
    }


def parse_classification_tags(tags: list[str]) -> dict[str, str]:
    """Extrai classificações do array de tags: {product_key: slug}.

    Ignora tags que não seguem o formato produto:perfil.
    """
    result: dict[str, str] = {}
    for tag in tags:
        if ":" not in tag:
            continue
        product_key, _, slug = tag.partition(":")
        if product_key in _CLASSIFICATION_VALID and slug in _CLASSIFICATION_VALID[product_key]:
            result[product_key] = slug
    return result


# Ordem de exibição dos produtos (usada no CSV)
PRODUCTS_ORDER: list[str] = [p["key"] for p in PRODUCTS]


def format_csv_columns(classifications: dict[str, str]) -> tuple[str, str]:
    """Formata as colunas 'produtos' e 'perfis' para CSV export.

    Retorna (produtos_label_str, perfis_label_str).
    """
    produto_labels: list[str] = []
    perfil_labels: list[str] = []
    for product_key in PRODUCTS_ORDER:
        slug = classifications.get(product_key)
        if slug is None:
            continue
        label = _PRODUCT_LABELS.get(product_key, product_key)
        profile_label = _PROFILE_LABELS.get(product_key, {}).get(slug, slug)
        produto_labels.append(label)
        perfil_labels.append(f"{label}: {profile_label}")
    return "; ".join(produto_labels), "; ".join(perfil_labels)

File: frontend/src/lib/taxonomy.ts
/**
 * Taxonomia CIMI Leads — cliente com cache e fallback hardcoded.
 *
 * Consome GET /api/taxonomy com stale-while-revalidate.
 * Se a rede falhar, usa o JSON embutido em ./taxonomy.json.
 *
 * IMPORTANTE: ./taxonomy.json é uma CÓPIA sincronizada de shared/taxonomy.json
 * (fonte editável). Editar só shared/taxonomy.json e rodar
 * scripts/sync_taxonomy.py — o pre-commit hook check-taxonomy-sync bloqueia
 * commit se divergir.
 */
import rawTaxonomy from "./taxonomy.json";

export interface TaxonomyProfile {
  slug: string;
  label: string;
}

export interface TaxonomyProduct {
  key: string;
  label: string;
  profiles: TaxonomyProfile[];
}

export interface Taxonomy {
  products: TaxonomyProduct[];
  legacy_profiles: Record<string, TaxonomyProfile[]>;
  interest_types: string[];
}

export const FALLBACK_TAXONOMY: Taxonomy = rawTaxonomy as Taxonomy;

let memoryCache: Taxonomy | null = null;
const LOCAL_KEY = "cimi_taxonomy_v1";
const TTL_MS = 5 * 60 * 1000; // 5 min

function _readLocal(): Taxonomy | null {
  if (typeof localStorage === "undefined") return null;
  try {
    const raw = localStorage.getItem(LOCAL_KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    if (parsed._fetchedAt && Date.now() - parsed._fetchedAt < TTL_MS) {
      return parsed.taxonomy as Taxonomy;
    }
  } catch {
    // ignore corrupt cache
  }
  return null;
}

function _writeLocal(t: Taxonomy): void {
  if (typeof localStorage === "undefined") return;
  try {
    localStorage.setItem(
      LOCAL_KEY,
      JSON.stringify({ taxonomy: t, _fetchedAt: Date.now() }),
    );
  } catch {
    // ignore quota errors
  }
}

function _timeoutSignal(ms: number): AbortSignal | undefined {
  if (
    typeof AbortSignal !== "undefined" &&
    typeof AbortSignal.timeout === "function"
  ) {
    return AbortSignal.timeout(ms);
  }
  return undefined;
}

/**
 * Busca taxonomia do backend com cache em memória + localStorage.
 * Stale-while-revalidate: retorna cache imediatamente e atualiza em background.
 */
export async function getTaxonomyCached(): Promise<Taxonomy> {
  if (memoryCache) return memoryCache;

  const local = _readLocal();
  if (local) {
    memoryCache = local;
    // revalidação em background (não esperamos)
    _revalidateBackground();
    return memoryCache;
  }

  return _fetchTaxonomy();
}

async function _fetchTaxonomy(): Promise<Taxonomy> {
  try {
    const res = await fetch("/api/taxonomy", {
      signal: _timeoutSignal(15_000),
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const taxonomy = (await res.json()) as Taxonomy;
    memoryCache = taxonomy;
    _writeLocal(taxonomy);
    return taxonomy;
  } catch {
    memoryCache = FALLBACK_TAXONOMY;
    return memoryCache;
  }
}

function _revalidateBackground(): void {
  _fetchTaxonomy().catch(() => {
    // silencioso em background
  });
}

/** Invalida cache (útil após deploy com taxonomia nova). */
export function invalidateTaxonomyCache(): void {
  memoryCache = null;
  if (typeof localStorage !== "undefined") {
    try {
      localStorage.removeItem(LOCAL_KEY);
    } catch {
      // ignore
    }
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Helpers de consulta (stateless — recebem Taxonomy)
// ═══════════════════════════════════════════════════════════════════════════════

export function getProductKeys(taxonomy: Taxonomy): string[] {
  return taxonomy.products.map((p) => p.key);
}

export function getProductLabel(taxonomy: Taxonomy, key: string): string | undefined {
  return taxonomy.products.find((p) => p.key === key)?.label;
}

export function getProfileLabel(
  taxonomy: Taxonomy,
  productKey: string,
  slug: string,
): string | undefined {
  const product = taxonomy.products.find((p) => p.key === productKey);
  const profile = product?.profiles.find((pr) => pr.slug === slug);
  if (profile) return profile.label;
  // legacy fallback
  const legacy = taxonomy.legacy_profiles[productKey]?.find((pr) => pr.slug === slug);
  return legacy?.label;
}

export function isValidProfile(
  taxonomy: Taxonomy,
  productKey: string,
  slug: string,
): boolean {
  const product = taxonomy.products.find((p) => p.key === productKey);
  if (!product) return false;
  return product.profiles.some((pr) => pr.slug === slug);
}

export function isLegacyProfile(
  taxonomy: Taxonomy,
  productKey: string,
  slug: string,
): boolean {
  const legacy = taxonomy.legacy_profiles[productKey];
  if (!legacy) return false;
  return legacy.some((pr) => pr.slug === slug);
}

/** Extrai classificações de tags: {product_key: slug} */
export function parseClassificationTags(
  taxonomy: Taxonomy,
  tags: string[],
): Record<string, string> {
  const known = new Set([
    ...taxonomy.products.map((p) => p.key),
    ...Object.keys(taxonomy.legacy_profiles),
  ]);
  const result: Record<string, string> = {};
  for (const tag of tags) {
    const idx = tag.indexOf(":");
    if (idx === -1) continue;
    const key = tag.slice(0, idx);
    const slug = tag.slice(idx + 1);
    if (known.has(key)) {
      result[key] = slug;
    }
  }
  return result;
}

/** Converte estado de classificação em tags prefixadas. */
export function classificacoesToTags(state: Record<string, string | null>): string[] {
  const tags: string[] = [];
  for (const [key, slug] of Object.entries(state)) {
    if (slug) tags.push(`${key}:${slug}`);
  }
  return tags;
}

/**
 * Converte tags em estado de classificação (todos os produtos conhecidos
 * presentes, com `null` para os não classificados).
 *
 * Aceita a taxonomia explicitamente (`tagsToClassificacoes(taxonomy, tags)`,
 * usada por quem já carregou via getTaxonomyCached) ou só as tags
 * (`tagsToClassificacoes(tags)`), caso em que usa FALLBACK_TAXONOMY — a
 * mesma fonte embutida (./taxonomy.json) usada como fallback de rede em
 * qualquer lugar do app.
 */
export function tagsToClassificacoes(
  tags: string[],
): Record<string, string | null>;
export function tagsToClassificacoes(
  taxonomy: Taxonomy,
  tags: string[],
): Record<string, string | null>;
export function tagsToClassificacoes(
  taxonomyOrTags: Taxonomy | string[],
  maybeTags?: string[],
): Record<string, string | null> {
  const taxonomy = Array.isArray(taxonomyOrTags) ? FALLBACK_TAXONOMY : taxonomyOrTags;
  const tags = Array.isArray(taxonomyOrTags) ? taxonomyOrTags : maybeTags ?? [];

  const state: Record<string, string | null> = {};
  for (const p of taxonomy.products) {
    state[p.key] = null;
  }
  const parsed = parseClassificationTags(taxonomy, tags);
  for (const [key, slug] of Object.entries(parsed)) {
    state[key] = slug;
  }
  return state;
}

/** Todas as tags de interesse aceitas (para UI de chips). */
export function getAllowedTags(taxonomy: Taxonomy): string[] {
  return taxonomy.interest_types;
}

/** Formata colunas de CSV (produtos e perfis). */
export function formatCsvColumns(
  classifications: Record<string, string>,
): { produtos: string; perfis: string } {
  const produtoLabels: string[] = [];
  const perfilLabels: string[] = [];
  for (const product of FALLBACK_TAXONOMY.products) {
    const slug = classifications[product.key];
    if (!slug) continue;
    const profile =
      product.profiles.find((p) => p.slug === slug) ??
      FALLBACK_TAXONOMY.legacy_profiles[product.key]?.find((p) => p.slug === slug);
    produtoLabels.push(product.label);
    perfilLabels.push(`${product.label}: ${profile?.label ?? slug}`);
  }
  return {
    produtos: produtoLabels.join("; "),
    perfis: perfilLabels.join("; "),
  };
}
